### ISSUE TITLE
Libraries' linking issues after cocoa-cb's changes (was: low quality jpeg screenshots after cocoa-cb's changes)

### ISSUE THREAD

**Issue by AirPort** [mpv-player/mpv#5528]
### mpv version and platform
mpv-git
macOs 10.13.3

### Reproduction steps
Set screenshot-format to jpg, screenshot-jpeg-quality to 100 and take a screenshot of a random video.

### Expected behavior
A high-quality jpeg screenshot is made.

### Actual behavior
jpeg screenshots take up less than half the space as before and are noticeably low quality. png screenshots seems unaffected and are as good as ever. Here some comparisons of the same frame:

git master, [png screen](https://0x0.st/sTT4.png)
git master, [jpeg screen](https://0x0.st/sTTJ.jpg)
f670c64 (last commit before the cocoa-cb changes), [jpeg screen](https://0x0.st/sTTy.jpg)

I also tried to take a [jpeg screen](https://0x0.st/sTTt.jpg) with git master and the deprecated cocoa backend, and it's even more strange: while some application, most notably firefox, shows it identical to low quality cocoa-cb screen, [this](https://0x0.st/sTTv.png) is what it looks like in Preview.app.

(I'm not sure if it matters, but all these tests are performed with an HiDPI screen).

### Log file
[git master](https://0x0.st/sTTl.txt)
[f670c64](https://0x0.st/sTTI.txt)

The main difference seems to be the
`[ffmpeg] mjpeg: intra_quant_bias = 96 inter_quant_bias = 0`
line in the git master log.

**Comment by Akemi** [comment]
isn't it more likely related to f17246fec15963e7cc098cf7ba7263ad02877b99 and using 9f595f3a80eeaa0bfbda5702031f054f684f6123?

**Comment by AirPort** [comment]
I also though that at first, but bisecting it I found out the issue started at c5e4538, before f17246f. Also the f670c64 version I reported on in the issue is still working and came out after 9f595f3.

**Comment by Akemi** [comment]
> I also tried to take a [jpeg screen](https://0x0.st/sTTt.jpg) with git master and the deprecated cocoa backend, and it's even more strange: while some application, most notably firefox, shows it identical to low quality cocoa-cb screen, [this](https://0x0.st/sTTv.png) is what it looks like in Preview.app.

that exact screen is butchered in preview?

it would probably also helpful to reproduce it with `--no-config` and the minimal settings you need to reproduce.

**Comment by AirPort** [comment]
>that exact screen is butchered in preview?

Yep.

>it would probably also helpful to reproduce it with --no-config and the minimal settings you need to reproduce.

Done with a barebones --no-config and no other option, still reproducible: https://0x0.st/sTT7.jpg.

**Comment by Akemi** [comment]
that exact screenshot shows just fine here in Preview.
![screen shot 2018-02-14 at 16 20 28](https://user-images.githubusercontent.com/680386/36211958-ff97d414-11a2-11e8-827f-5c35c32c0c43.png)

**Comment by AirPort** [comment]
Tried a different host, redownloading from it shows the garbled pattern: https://my.mixtape.moe/mejcyr.jpg

(This one could also be a bug in preview, as the pattern changes slightly everytime I open the image, even though it's the first time I see such a thing happening).

**Comment by Akemi** [comment]
still the same, it works. both images also have the same CRC for me when downloading.

**Comment by Akemi** [comment]
i noticed on your git master log you have `jpeg` enabled but not on the f670c64 log.

**Comment by Akemi** [comment]
sry i switched the logs accidentally. i would say the problem is because one version uses ffmpeg to write the jpeg the other uses a dedicated lib.

**Comment by AirPort** [comment]
Oh damn, I think you're right. It always autodetected the library without specifying an option, so I didn't think to check that.
I'll try ASAP and see (I hope it's  not another detection issue like the ffmpeg's dependencies one).

**Comment by AirPort** [comment]
Yeah, as I suspected waf can't find libjpeg anymore (installed by brew), not even with explicit --enable-jpeg. Can you confirm it works for you?

**Comment by AirPort** [comment]
Ok, I tried to delete [this if](https://github.com/mpv-player/mpv/blob/master/waftools/checks/custom.py#L106) like you suggested in a comment in the cocoa-cb PR, and surprisingly now the same configure string that failed before (./waf configure --enable-zsh-comp --enable-libarchive --enable-libbluray --enable-dvdnav --enable-dvdread --enable-uchardet --enable-javascript --enable-lua --enable-lcms2 --enable-html-build --enable-libmpv-shared --enable-jpeg), now worked flawlessly, libjpeg detection included.

**Comment by Akemi** [comment]
hm maybe it would be better then to add that linking flag after the checks and when building.

**Comment by Akemi** [comment]
does this fix the dependency issue?

```
diff --git a/waftools/checks/custom.py b/waftools/checks/custom.py
index 14e59f3..6987424 100644
--- a/waftools/checks/custom.py
+++ b/waftools/checks/custom.py
@@ -100,15 +100,7 @@ def check_cocoa(ctx, dependency_identifier):
         includes         = ctx.srcnode.abspath(),
         linkflags        = '-fobjc-arc')
 
-    res = fn(ctx, dependency_identifier)
-    # on macOS we explicitly need to set the SDK path, otherwise it can lead to
-    # linking warnings or errors
-    if res:
-        ctx.env.append_value('LINKFLAGS', [
-            '-isysroot', ctx.env.MACOS_SDK
-        ])
-
-    return res
+    return fn(ctx, dependency_identifier)
 
 def check_openal(ctx, dependency_identifier):
     checks = [
diff --git a/wscript_build.py b/wscript_build.py
index 4627a9f..7a860c2 100644
--- a/wscript_build.py
+++ b/wscript_build.py
@@ -152,6 +152,13 @@ def build(ctx):
                               bridge, header, tgt, src)
         return task.exec_command(cmd)
 
+    if ctx.dependency_satisfied('macos-cocoa-cb') and ctx.env.MACOS_SDK:
+        # on macOS we explicitly need to set the SDK path, otherwise it can lead to
+        # linking warnings or errors
+        ctx.env.append_value('LINKFLAGS', [
+            '-isysroot', ctx.env.MACOS_SDK
+        ])
+
     if ctx.dependency_satisfied('macos-cocoa-cb'):
         swift_source = [
             ( "video/out/cocoa_cb_common.swift" ),

```

**Comment by AirPort** [comment]
Yep, seems like it!

**Comment by Akemi** [comment]
added the fix to #5529.

**Comment by AirPort** [comment]
Tried it again with the PR, I can confirm everything seems to be working now. Thank you!
